File: metro.js
// metro.js
const fs = require('fs');

let adj = {};
let stationLines = {};

const key = (s,l)=>`${s}|${l}`;

function addEdge(u,v){
    if(!adj[u]) adj[u]=[];
    adj[u].push(v);
}

function buildGraph(data){
    for(const station in data){
        const lines = data[station].lines;

        for(const line in lines){
            if(!stationLines[station]) stationLines[station]=[];
            stationLines[station].push(line);

            const conn = lines[line];

            if(conn.prev)
                addEdge(key(station,line),
                        key(conn.prev,line));
            if(conn.next)
                addEdge(key(station,line),
                        key(conn.next,line));
        }
    }

    for(const st in stationLines){
        for(const l1 of stationLines[st])
            for(const l2 of stationLines[st])
                if(l1!==l2)
                    addEdge(key(st,l1), key(st,l2));
    }
}

const data = JSON.parse(fs.readFileSync("dmrc.json"));
buildGraph(data);
console.log("Graph ready");

File: metro.cs
// Metro.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using System.IO;

class Metro {
    static Dictionary<string, List<string>> adj = new();
    static Dictionary<string, List<string>> stationLines = new();

    static string Key(string s,string l)=> $"{s}|{l}";

    static void AddEdge(string u,string v){
        if(!adj.ContainsKey(u)) adj[u]=new();
        adj[u].Add(v);
    }

    static void BuildGraph(JObject data){
        foreach(var station in data){
            var lines = (JObject)station.Value["lines"];

            foreach(var line in lines){
                stationLines.TryAdd(station.Key,new());
                stationLines[station.Key].Add(line.Key);

                var conn = (JObject)line.Value;

                if(conn["prev"]!=null)
                    AddEdge(Key(station.Key,line.Key),
                            Key(conn["prev"].ToString(),line.Key));

                if(conn["next"]!=null)
                    AddEdge(Key(station.Key,line.Key),
                            Key(conn["next"].ToString(),line.Key));
            }
        }
    }

    static void Main(){
        var json = File.ReadAllText("dmrc.json");
        var data = JObject.Parse(json);
        BuildGraph(data);
        Console.WriteLine("Graph built");
    }
}
